mirgecom/steppers.py

```diff
@@ -93,3 +96,193 @@ def advance_state(rhs, timestepper, checkpoint, get_timestep,
+    component_id
+        State id (required input for leap method generation)
+    t: float
+        Time at which to start
+    istep: int
+        Step number from which to start
+
+    Returns
+    -------
+    istep: int
+        the current step number
+    t: float
+        the current time
+    state: numpy.ndarray
+    """
+    if t_final <= t:
+        return istep, t, state
+
+    # Generate code for Leap method.
+    dt = get_timestep(state=state)
+    stepper_cls = generate_singlerate_leap_advancer(timestepper, component_id,
+                                                    rhs, t, dt, state)
+    while t < t_final:
+
+        if logmgr:
+            logmgr.tick_before()
+
+        dt = get_timestep(state=state)
+        if dt < 0:
+            return istep, t, state
+
+        checkpoint(state=state, step=istep, t=t, dt=dt)
+
+        # Leap interface here is *a bit* different.
+        for event in stepper_cls.run(t_end=t+dt):
+            if isinstance(event, stepper_cls.StateComputed):
+                state = event.state_component
+                t += dt
+                istep += 1
+                if logmgr:
+                    set_dt(logmgr, dt)
+                    set_sim_state(logmgr, dim, state, eos)
+                    logmgr.tick_after()
+
+    return istep, t, state
+
+
+def generate_singlerate_leap_advancer(timestepper, component_id, rhs, t, dt,
+                                      state):
+    """Generate single-rate Leap code to be used to advance state.
+
+    Parameters
+    ----------
+    timestepper
+        Leap method that advances the state from t=time to t=(time+dt), and
+        returns the advanced state.
+    component_id
+        State id (required input for leap method generation)
+    rhs
+        Function that should return the time derivative of the state.
+        This function should take time and state as arguments, with
+        a call looking like rhs(t, state).
+    t: float
+        Time at which to start
+    dt: float
+        Initial timestep to be set by leap method
+    state: numpy.ndarray
+        Agglomerated object array containing at least the state variables that
+        will be advanced by this stepper
+
+    Returns
+    -------
+    dagrt.codegen.python.StepperInterface
+        Python class implementing leap method, and generated by dagrt
+    """
+    code = timestepper.generate()
+    from dagrt.codegen import PythonCodeGenerator
+    codegen = PythonCodeGenerator(class_name="Method")
+    stepper_cls = codegen.get_class(code)(function_map={
+        "<func>" + component_id: rhs,
+        })
+    stepper_cls.set_up(t_start=t, dt_start=dt, context={component_id: state})
+
+    return stepper_cls
+
+
+def advance_state(rhs, timestepper, checkpoint, get_timestep, state, t_final,
+                    component_id="state", t=0.0, istep=0, logmgr=None,
+                    eos=None, dim=None):
+    """Determine what stepper we're using and call the appropriate advancer.
```

Comment:
> call the appropriate advancer.

This isn't useful. The functions called by this aren't included in the docs (purposefully so: they're not public interfaces), so this function needs to document what happens.